### PR TITLE
Migration to new IAdjudicator using bool

### PR DESCRIPTION
**Pull request by mod**
Status is moving into Custody implementation

### CODE REVIEW

File: contract/src/adjudicators/Counter.sol [hunk 1]
```diff
@@ -24,7 +24,8 @@ contract Counter is IAdjudicator {
     /// @notice Error thrown when insufficient signatures are provided
     error InsufficientSignatures();
 
-    uint256 private constant FINAL_COUNTER = 1000;
+    // The custody contract now determines if a state is final based on signatures
+    // not on reaching a specific counter value
     uint256 private constant HOST = 0;
     uint256 private constant GUEST = 1;
```
Comment (mod) [inline contract/src/adjudicators/Counter.sol:31]: CounterApp {
  uint256 value; // Participant agree on the starting value
  uint256 final; // Participant agree on the final count
  uint256 version; // Each version of CounterApp increment version number from 0  
}

File: contract/src/Custody.sol [hunk 2]
```diff
@@ -373,8 +422,15 @@ contract Custody is IChannel, IDeposit {
         uint256 correctedAmount = channelBalance > amount ? amount : channelBalance;
         meta.tokenBalances[token] -= correctedAmount;
 
-        Account storage accountInfo = _accounts[account];
-        accountInfo.tokens[token].locked -= correctedAmount;
-        accountInfo.tokens[token].available += correctedAmount;
+        Ledger storage accountInfo = _accounts[account];
+
+        // Check locked amount before subtracting to prevent underflow
```
Comment (nksazonov) [inline contract/src/Custody.sol:427]: I think this logic is redundant as this situation should not be possible. Still, if this "safeguard" is a requirement, let's move the line `accountInfo.tokens[token].available += amountToUnlock` out from the `if` statement, as channel balance (`meta.tokenBalances[token] -= correctedAmount`) is decreased regardless, so available balance should increase regardless too.

File: contract/src/adjudicators/Counter.sol [hunk 3]
```diff
@@ -24,99 +24,86 @@ contract Counter is IAdjudicator {
...
+        // INITIAL STATE: version 0 requires both signatures.
+        if (candidateState.version == 0) {
+            if (
+                !Utils.verifySignature(stateHash, candidate.sigs[0], chan.participants[HOST])
+                    || !Utils.verifySignature(stateHash, candidate.sigs[1], chan.participants[GUEST])
+            ) {
+                return false;
             }
+            // Valid initial state.
+            return true;
```
Comment (nksazonov) [inline contract/src/adjudicators/Counter.sol:75]: This can be simplified to `return Utils.verifySignature(...) && Utils.verifySignature(...);`

Comment (nksazonov) [reply contract/src/adjudicators/Counter.sol]: c3a15f89e51784a39abaa078c81481a3aa8a1b7f

File: contract/src/Custody.sol [hunk 4]
```diff
@@ -35,23 +35,25 @@ contract Custody is IChannel, IDeposit {
...
-    struct Account {
-        mapping(address token => Funds funds) tokens; // Token balances
+    struct Ledger {
+        mapping(address token => Account funds) tokens; // Token balances
         EnumerableSet.Bytes32Set channels; // Set of user ChannelId
     }
 
     mapping(bytes32 channelId => Metadata chMeta) internal _channels;
-    mapping(address account => Account) internal _accounts;
+    mapping(address account => Ledger) internal _accounts;
```
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol:56]: good idea

File: contract/src/Custody.sol [hunk 5]
```diff
@@ -61,12 +63,11 @@ contract Custody is IChannel, IDeposit {
             bool success = IERC20(token).transferFrom(account, address(this), amount);
             require(success, TransferFailed(token, address(this), amount));
         }
-
         _accounts[msg.sender].tokens[token].available += amount;
     }
 
     function withdraw(address token, uint256 amount) external {
-        Account storage account = _accounts[msg.sender];
+        Ledger storage account = _accounts[msg.sender];
```
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol:70]: yes, ok

File: contract/src/Custody.sol [hunk 6]
```diff
@@ -184,18 +218,28 @@ contract Custody is IChannel, IDeposit {
...
+            revert InvalidState();
+        }
 
-        if (status == IAdjudicator.Status.ACTIVE || status == IAdjudicator.Status.PARTIAL) {
-            // Valid challenge, start challenge period
+        // Validate the candidate state using adjudicator
+        // Adjudicators only validate state transitions, not channel status
+        bool isValid = _validateState(meta, candidate, proofs);
+
+        // FIXME: challenge can only be valid if it's more recent state than the checkpointed state
```
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol]: define the interface, but this could be very useful, maybe it should be native in the IAdjudicator IComparable (maybe it exists already)

what would be the meaning of 0 ? same version ?

File: contract/src/Custody.sol [hunk 7]
```diff
@@ -209,14 +253,25 @@ contract Custody is IChannel, IDeposit {
     function checkpoint(bytes32 channelId, State calldata candidate, State[] calldata proofs) external {
         Metadata storage meta = _requireValidChannel(channelId);
 
+        // Only allow checkpoints on funded channels
```
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol:280]: good point, you can create an annotation

File: contract/src/Custody.sol [hunk 8]
```diff
@@ -298,44 +357,34 @@ contract Custody is IChannel, IDeposit {
...
-        IAdjudicator adjudicator = IAdjudicator(ch.adjudicator);
-        // Convert to calldata by passing individual parameters
-        return adjudicator.adjudicate(ch, candidate, proofs);
+                return valid;
+            }
+
+            return valid;
+        } catch {
+            // If the adjudicator call reverts, treat as invalid state
+            meta.status = Status.INVALID;
```
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol:413]: this method is outdated, and should not be used. use _validateState instead

File: contract/src/Custody.sol (hunk 2 above)
Comment (mod) [reply contract/src/Custody.sol:455]: Ok, we will have to test well those method on edge cases

File: contract/src/Custody.sol [hunk 9]
```diff
@@ -143,31 +151,36 @@ contract Custody is IChannel, IDeposit {
...
+            // Verify stateHash is signed by both participants
+            bytes32 stateHash = Utils.getStateHash(ch, depositState);
+
+            // Verify HOST signature
+            bool hostSignatureValid = Utils.verifySignature(stateHash, depositState.sigs[0], ch.participants[HOST]);
+            require(hostSignatureValid, "Host signature invalid");
+
+            // Verify GUEST signature
+            bool guestSignatureValid = Utils.verifySignature(stateHash, depositState.sigs[1], ch.participants[GUEST]);
+            require(guestSignatureValid, "Guest signature invalid");
```
Comment (nksazonov) [inline contract/src/Custody.sol:169]: This can be substituted with `_verifyAllSignatures(...)`

**Review by nksazonov — APPROVED** [review]

File: contract/src/Custody.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply contract/src/Custody.sol:413]: My comment points to `_validateState` method, which saves the `INVALID` state on-chain. What is the point of that? Wouldn't it be better to revert if a state is invalid, so that we don't have something invalid on-chain?

File: contract/src/Custody.sol (hunk 9 above)
Comment (nksazonov) [reply contract/src/Custody.sol]: 90d201e7b0cd953cd9cc6c659c53176262a2e090

File: contract/src/Custody.sol (hunk 2 above)
Comment (nksazonov) [reply contract/src/Custody.sol:449]: 90d201e7b0cd953cd9cc6c659c53176262a2e090

File: contract/src/Custody.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply contract/src/Custody.sol]: 90d201e7b0cd953cd9cc6c659c53176262a2e090

File: contract/src/Custody.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply contract/src/Custody.sol]: 90d201e7b0cd953cd9cc6c659c53176262a2e090

File: contract/src/Custody.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply contract/src/Custody.sol:275]: c18d357f68a63fbea81c80716b94daaed48a815e

File: contract/src/Custody.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (mod) [reply contract/src/Custody.sol:408]: Just add a fixme in the code, I will get back to this

File: contract/src/Custody.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (nksazonov) [reply contract/src/Custody.sol]: yeah, `0` is for `equal` states.
